box-content-sdk/src/main/java/com/box/androidsdk/content/requests/BoxRequestsFile.java

```diff
@@ -1206,4 +1219,443 @@ public void setRequestedPage(int page) {
+        static final String DIGEST_HEADER = "digest";
+        private static final String CONTENT_RANGE_HEADER = "content-range";
+        static final String DIGEST_HEADER_PREFIX_SHA = "sha=";
+
+        private final int mPartNumber;
+        private File mFile;
+        private int mCurrentChunkSize;
+        private final BoxUploadSession mUploadSession;
+
+        /**
+         *
+         * @param file Creates an upload file session part  with the default parameters
+         * @param uploadSession
+         * @param partNumber
+         * @param session
+         */
+        public UploadSessionPart(File file, BoxUploadSession uploadSession, int partNumber, BoxSession session) throws IOException {
+            super(BoxUploadSessionPart.class, uploadSession.getEndpoints().getUploadPartEndpoint(), session);
+            mRequestUrlString = uploadSession.getEndpoints().getUploadPartEndpoint();
+            mRequestMethod = Methods.PUT;
+            mPartNumber = partNumber;
+            mUploadSession = uploadSession;
+            mFile = file;
+            mCurrentChunkSize = BoxUploadSession.getChunkSize(uploadSession, partNumber, mFile.length());
+            mContentType = ContentTypes.APPLICATION_OCTET_STREAM;
+
+        }
+
+
+
+        @Override
+        protected void createHeaderMap() {
+            super.createHeaderMap();
+            long offset = mPartNumber * mUploadSession.getPartSize();
+            //Content-Range: bytes offset-part/totalSize
+
+            long lastByte = offset + mCurrentChunkSize - 1;
+            mHeaderMap.put(CONTENT_RANGE_HEADER,
+                    "bytes " + offset + "-" + lastByte + "/" + mFile.length());
+            mHeaderMap.put(DIGEST_HEADER, DIGEST_HEADER_PREFIX_SHA + mUploadSession.getFieldPartsSha1().get(mPartNumber));
+        }
+
+
+        @Override
+        protected void setBody(BoxHttpRequest request) throws IOException  {
+            FileInputStream inputStream = new FileInputStream(mFile);
+            //skip previous parts
+            inputStream.skip(mPartNumber * mUploadSession.getPartSize());
+
+            //Write bytes to URLConnection of the request
+            URLConnection urlConnection = request.getUrlConnection();
+            urlConnection.setDoOutput(true);
+            OutputStream output = urlConnection.getOutputStream();
+            byte[] byteBuf = new byte[SdkUtils.BUFFER_SIZE];
+            int totalBytesRead = 0;
+            try {
+                if (inputStream.available() > 0) {
+                    int bytesRead = 0;
+                    while (totalBytesRead < mCurrentChunkSize && inputStream.available() > 0) {
+                        if (Thread.currentThread().isInterrupted()) {
+                            InterruptedException e = new InterruptedException();
+                            throw e;
+                        }
+                        bytesRead = inputStream.read(byteBuf);
```

Comment:
In this moment, in case we are in the last byteBuf chunk of the part, we have a high possibility that we read more than the chunk total size.
To fix that, I suggest you check the part remaining bytes, and make sure that it is bigger than the byteBuf size. Otherwise, make sure to read (and write to output) only the remaining bytes.

Reply:
It's actually okay right now..given chunk size is a multiple of buffer size; but good to protect against future change.